Frequent Itemset Clustering (Apriori and ECLAT)

Wander03 opened the pull request:
@kbodwin

- Relates to other conversations about column-based clustering, e.g. #66

- Adds a partition mode with engine arules to tidyclust (freq_itemsets)
- Adds custom cluster and predict functions for `freq_itemsets()`
- Adds `extract_predictions()` which reformates `predict()` output into a more readable format
- Adds `augment_itemset_predict()` which reformates `predict()` output for metric functions (e.g. in yardstick)

Note:
- `devtools::check()` resulted in a warning about code dependencies from purr and stringr

Wander03 committed:
created bsaed function for frequent itemsets and association rules

Wander03 committed:
testing new functions

Wander03 committed:
clustering for freq itemsets function

Wander03 committed:
progress!

Wander03 committed:
fix conditions

Wander03 committed:
fix conditions

Wander03 committed:
Update text

Wander03 committed:
change method to mining_method

Wander03 committed:
create vignette for freq itemsets

Wander03 committed:
bug fixing

Wander03 committed:
fixed name

Wander03 committed:
premptive changes

Wander03 committed:
bug fixing freq itemsets

Wander03 committed:
code formatting

Wander03 committed:
bug fixes

Wander03 committed:
updating cluster functions

Wander03 committed:
save average supports for each cluster (to be used in predict)

Wander03 committed:
predict not saving output

Wander03 committed:
some change

Wander03 committed:
fixed predcit! Proba is now put in N/A spots

Wander03 committed:
remove avg support tracker (unused)

Wander03 committed:
change best cluster to prioritize size then support

add predict to vingette

Wander03 committed:
vignette testing

Wander03 committed:
predict output formated & cutoff implemented

Wander03 committed:
create holder for extract_predictions function (placeholder name)

Wander03 committed:
hard code cutoff

Wander03 committed:
change predict formating

Wander03 committed:
something>

Wander03 committed:
extract_predictions complete! (still needs a better name)

Wander03 committed:
move detail text

Wander03 committed:
add tuning for min_support

Wander03 committed:
tune update

Wander03 committed:
improved example code

Wander03 committed:
fix params help doc

Wander03 committed:
create test files (TODO: create tests)

Wander03 committed:
adding raw to predict

Wander03 committed:
split regualar predict and raw predict.
NOTE: For the normal predict it is what we want, nested data frames in a data frame. However, the structure of the raw output is a list with data frames).

Wander03 committed:
predict fixed! Output is the same :D

Wander03 committed:
change default to eclat

Wander03 committed:
fixing replacing wrong part from earlier commit

Wander03 committed:
updating with correct default

Wander03 committed:
augment code written, add correct header text and move unecessary code

Wander03 committed:
standardize column names

Wander03 committed:
testing

Wander03 committed:
testing2

Wander03 committed:
hide predict dataframe from arules::inspect()

Wander03 committed:
remove `` from predict output item names

Wander03 committed:
added note

Wander03 committed:
re-roder doesnt matter for fit

Wander03 committed:
hide freq itemset output from auto displaying when extracting cluster assignment

Wander03 committed:
rename col name in predict

Wander03 committed:
vignettes update with new info from thesis

Wander03 committed:
added header descriptions about functions

Wander03 committed:
added convergence limit and warning message

Wander03 committed:
update freq_itemsets extract_fit_summary and ? information

Wander03 committed:
vignette update

Wander03 committed:
create test cases for freq_itemsets

Wander03 committed:
move min_support tuning to dials

Wander03 committed:
Merged upstream/main into main

Wander03 committed:
re-ran test cases

Wander03 committed:
remove assoc_rules

EmilHvitfeldt commented on tests/testthat/test-extract_centroids.R:
```diff
@@ -1,3 +1,11 @@
+toy_df <- data.frame(
+  'beer'    = c(F, T, T, T, F),
+  'milk'    = c(T, F, T, T, T),
+  'bread'   = c(T, T, F, T, T),
+  'diapers' = c(T, T, T, T, T),
+  'eggs'    = c(F, T, F, F, F)
+)
```
This does two things, stops the usage of `'` over `"` and uses the full name for `TRUE` and `FALSE`

EmilHvitfeldt commented on R/augment_itemset_predict.R:
```diff
@@ -0,0 +1,119 @@
+#'     \item `truth`: The true value for the item from `truth_output`.
+#'   }
+#'   This output is suitable for direct use with `yardstick` metric functions.
+#' @export
+
+augment_itemset_predict <- function(pred_output, truth_output) {
```
All exported functions need examples.

I would also like to see the example to help determine the use of it

EmilHvitfeldt commented on tests/testthat/test-extract_centroids.R:
```diff
@@ -64,3 +72,12 @@ test_that("prefix is passed in extract_centroids()", {
 })
+
+test_that("extract_centroids errors for freq_itemsets", {
+  set.seed(1234)
```
please add `skip_if_not_installed("arules")` to all tests that use `freq_itemsets()`

EmilHvitfeldt commented on tests/testthat/test-freq_itemsets-arules.R:
```diff
@@ -0,0 +1,75 @@
+  set.seed(1234)
+  fi_fit <- freq_itemsets(min_support = 0.5) %>%
+    set_engine("arules") %>%
+    fit(~., toy_df %>% dplyr::mutate(across(everything(), as.numeric)))
```
i see that sometimes `dplyr::mutate(across(everything(), as.numeric))` is called and sometimes it isn't. is there a reason why?

Wander03 replied:
The values in the data frame can either be TRUE/FALSE or 1/0, however if they are TRUE/FALSE then these values get tacked onto the end of the column names (e.g. beerTRUE) when inspecting the frequent itemsets (e.g. arules::DATAFRAME(fi_fit$fit)). But, this does not occur when using the arules functions directly for some reason, so for ensure that the text comparisons are exactly the same when looking at these I made sure to use 1/0 rather than TRUE/FALSE.

EmilHvitfeldt commented on R/augment_itemset_predict.R:
```diff
@@ -0,0 +1,119 @@
+  # Extract all predictions (bind all .pred_cluster dataframes)
+  preds_df <- dplyr::bind_rows(pred_output$.pred_cluster, .id = "row_id") %>%
+    dplyr::filter(!is.na(.pred_item)) %>%  # Keep only rows with predictions
+    dplyr::mutate(item = stringr::str_remove_all(item, "`")) %>% # Remove backticks from item names
```
There are a couple of stringr functions used here. I would prefer if we used base-R functions instead https://stringr.tidyverse.org/articles/from-base.html

EmilHvitfeldt commented on R/extract_cluster_assignment.R:
```diff
@@ -159,6 +159,79 @@ extract_cluster_assignment.hclust <- function(
+  items <- attr(object, "item_names")
+  itemsets <- arules::DATAFRAME(object)
+
+  itemset_list <- lapply(strsplit(gsub("[{}]", "", itemsets$items), ","), stringr::str_trim)
```
more stringr https://stringr.tidyverse.org/articles/from-base.html

EmilHvitfeldt commented on R/predict_helpers.R:
```diff
@@ -177,3 +177,88 @@ make_predictions <- function(x, prefix, n_clusters) {
+  # Extract frequent itemsets and their supports
+  items <- attr(object, "item_names")
+  itemsets <- arules::DATAFRAME(object)
+  frequent_itemsets <- lapply(strsplit(gsub("[{}]", "", itemsets$items), ","), stringr::str_trim)
```
more stringr https://stringr.tidyverse.org/articles/from-base.html

EmilHvitfeldt commented on R/predict_helpers.R:
```diff
@@ -177,3 +177,88 @@ make_predictions <- function(x, prefix, n_clusters) {
+
+    # Create result data frame
+    data.frame(
+      item = stringr::str_remove_all(items, "`"), # Remove backticks from item names
```
more stringr https://stringr.tidyverse.org/articles/from-base.html

EmilHvitfeldt commented on R/extract_predictions.R:
```diff
@@ -0,0 +1,30 @@
+
+  # Process each observation and combine results using reduce
+  result_df <- data_frames %>%
+    purrr::reduce(.f = ~ {
```
please use the `reduce()` from `compat-purrr.R`

EmilHvitfeldt commented on R/extract_cluster_assignment.R:
```diff
@@ -173,3 +246,34 @@ cluster_assignment_tibble <- function(
+  unique_non_zero_clusters <- unique(non_zero_clusters)
+
+  # Map each unique non-zero cluster to a new cluster starting from Cluster_1
+  cluster_map <- setNames(paste0(prefix, seq_along(unique_non_zero_clusters)), unique_non_zero_clusters)
```
```suggestion
  cluster_map <- stats::setNames(paste0(prefix, seq_along(unique_non_zero_clusters)), unique_non_zero_clusters)
```

EmilHvitfeldt reviewed (commented):
To more things:

- Add the following `.pred_item item preds row_id setNames truth_value` to `utils::globalVariables()` in `aaa.R`.
- Add exported functions to _pkgdown.yml`

i think i would like to chat about these prediction types in https://github.com/tidymodels/tidyclust/issues/211 before going through with this PR

Wander03 committed:
Add the following

.pred_item
item
preds
row_id
setNames
truth_value

to utils::globalVariables() in aaa.R

Wander03 committed:
rename `extract_predictions` to `extract_itemset_predictions`

add example to `extract_itemset_predictions`

Wander03 committed:
rename `extract_predictions` to `extract_itemset_predictions`

Wander03 committed:
Add exported functions to _pkgdown.yml

Wander03 committed:
convert all rlang::abort() calls to use {cli}

Wander03 committed:
edit toy_df and toy_pred to use " instead of ' and TRUE/FALSE instead of T/F

Wander03 committed:
add example to `augment_itemset_predict`

Wander03 committed:
add skip_if_not_installed("arules") to all tests that use freq_itemsets()

Wander03 committed:
use base R rather than stringr

Wander03 committed:
use the reduce() from compat-purrr.R

Wander03 committed:
stats::setNames

Wander03 commented:
Hi Emil! I believe that I addressed all your comments, please let me know if I missed something or if there is something else I need to edit.
